### ISSUE TITLE
[Feature Request] New shape modifier indicating "zero or one" axis

### ISSUE THREAD

**Issue by DecAngel** [google/jaxtyping#99]
Hi,

I was wondering if there is a way to annotate "zero or one" axis (the axis can optionally exist). I only find '\*' in the documentation that indicates zero or more axes, allowing more than one axis to exist. Like '\*' in regular expression, I think '?' can be introduced to indicate "zero or one" axis.

I was hoping to annotate functions like this.

```python
# the '?' modifier matches zero or one axis
def my_collate_fn(xs: List[Float[Tensor, '?batchsize channel height width']]) -> Float[Tensor, 'newbatchsize channel height width']:
    if xs[0].ndim == 3:
        return torch.stack(xs, dim=0)
    elif xs[0].ndim == 4:
        return torch.cat(xs, dim=0)
    else:
        raise ValueError('number of axes should be 3 or 4')
```

Any help is appreciated. Thanks in advance!

**Comment by patrick-kidger** [comment]
You can express this as a `typing.Union` of the two shapes.
Generally speaking I'd regard this kind of rank-polymorphism as an antipattern, so I don't want to add special syntax for it.

**Comment by DecAngel** [comment]
Thanks, I wasn't aware that this case could be covered by Union.
